### ISSUE TITLE
Implement Integration Tests for Enhanced Reliability

### ISSUE THREAD

**Issue by maxlambrecht** [spiffe/spire-controller-manager#309]
To ensure the SPIRE Controller Manager operates reliably, this issue proposes the implementation of a suite of integration tests. Adapting the existing demo to run automatically serves as an excellent starting point. This approach will help systematically validate the system's functionality, laying the foundation for a robust testing framework.

**Comment by kfox1111** [comment]
In a way, we're doing some of that already in the helm-chart-hardened repo as well. Would it make sense to team up between projects?

**Comment by maxlambrecht** [comment]
I'm customizing the Dockerfile to incorporate specific internal images for our use case. To ensure these adjustments don't inadvertently cause regressions, having integration tests within the same repository would be extremely useful. This approach avoids the complexity of relying on another repo, streamlining the testing process.

**Comment by kfox1111** [comment]
Fair. Depends on how far you want to take the tests i guess. Could introduce complexities of trying to integrate it with all the other bits by yourself. Like, does it work properly when configured to deploy a working mysql with k8s -> spire-controller-manager -> spire -> agent syncing the dnsname to allow authentication. We're already doing some of those type of tests, and being able to 'helm install spire --set spire-server.controllerManager.image=customimage  may be relatively easy.

**Comment by maxlambrecht** [comment]
Indeed, the tests you've described seem well-suited for the helm charts repository, focusing on end-to-end functionality including deployment and interaction with Kubernetes. For the `spire-controller-manager`, I'm advocating for integration tests that concentrate on the core functionality — particularly, the creation and verification of entries in the spire-server under diverse controller manager configurations.

**Comment by azdagron** [comment]
This is very much desired and a prerequisite for the project seeking the "Production" badge from the SPIFFE organization.

While it would be nice to not have repeated effort, I feel like the integration tests for this repo should be independent of the helm chart and target specific features to the controller.
